imaids/utils.py

```diff
@@ -131,6 +136,62 @@ def cosine_function(z, bamp, freq, phase):
     return bamp*_np.cos(freq*z + phase)
 
 
+def get_beff_from_model(model, period, polarization, hmax, x):
+    """Calculate effective field amplitude from model.
+
+    Args:
+        model (radia object): ID model.
+        period (float): ID period.
+        polarization (string): polarization of radiation. (['hp', 'vp', 'cp'])
+        hmax (int): max field harmonic to be considered.
+        x (float): horizontal position to calc field [mm]
+
+    Returns:
+        float : effective field amplitude [T]
+        float: first harmonic field amplitude [T]
+        numpy.ndarray: field along longitudinal positions [T]
+    """
+    zmin = -2*period
+    zmax = 2*period
+    npts = 201
+    z = _np.linspace(zmin, zmax, npts)
+    bvec = model.get_field(x=x,z=z)
+    if polarization == 'vp':
+        b = bvec[:, 0]
+    elif polarization in ('hp', 'cp'):
+        b = bvec[:, 1]
+
+    freq0 = 2*_np.pi/period
+    hs = _np.array(range(1, hmax+1, 2))
+    freqs = hs*freq0
+    amps,*_ = fit_fourier_components(b, freqs, z)
+    beff = _np.sqrt(_np.sum(_np.power(amps/hs, 2)))
+    return beff, amps[0], b
+
+
+def get_beff_fit(gap_over_period, beff, br):
+    """Return fitted B(Gap/period) exponential curve parameters."""
+    a0, b0, c0 = 2, -3, 0
+
+    def fit_function(x, a, b, c):
+        return br*a*_np.exp(b*x + c*(x**2))
+
+    opt = _optimize.curve_fit(
+        fit_function, gap_over_period, beff, p0=(a0, b0, c0))[0]
+
+    return opt
+
+
+def undulator_b_to_k(b, period):
+    """Field amplitude to K conversion."""
+    return ELEMENTARY_CHARGE * period * b / (2 * _np.pi * ELECTRON_MASS * LIGHT_SPEED)
+
+
+def undulator_k_to_b(k, period):
+    """K to field amplitude conversion."""
+    return (2 * _np.pi * ELECTRON_MASS * LIGHT_SPEED * k) / (ELEMENTARY_CHARGE * period)
+
```

Comment:
At the `SinusoidalFieldSource` there is also `calc_deflection_parameter` method, but there is no reverse method, such as the one you implemented. I believe the reverse method (k to b) could be moved to there.
Maybe there it could be named as something like "calc_b_from_deflection_parameter' or something like this.

Reply:
Hi Lucas! The idea of implementing this function was also to convert field amplitudes given by files, it could be field measurements or general saved data. We could do it using SinusoidalFieldSource too, but it would be necessary to create an object to execute this function, right?

Reply:
You mean you are planning to add the files feature later? Sorry, I still didn't get how this one is different from the already existing one... :no_mouth: :thinking:

Reply:
I think the differences between these two functions are the following:

calc_deflection_parameter is a method of SinusoidalFieldSource class, which is a base class for InsertionDeviceModel, so if we want to know the deflection parameter of an ID that has already a Radia model, we use the method calc_deflection_parameter.

undulator_b_to_k is not a method, so even if we don't have an ID Radia model, but we have a file with field data we can obtain the deflection parameter, to do this we only need to call the function fit_fourier_components that will return the maximum value of the field data and then apply this result as an argument of undulator_b_to_k.

Reply:
Ok! I've got it now.

One last comment. The `SinusoidalFieldSource` class is also base to a class called InsertionDeviceData. If you create an object of such class you can import data to it and this object will act just like if it were a model object. You can even use get_field() on it and the field map will be interpolated. Do you think this could match your use case?

If not, or if would like to have a simple callable function for the deflection parameter... like a "ready-to-go calculation shortcut", it would be fine to have it on utils.py

In this case, I would just suggest this new `undulator_b_to_k` to be called inside `calc_deflection_parameter` to perform the calculation there.  This would avoid having two independent codes doing the same calculation (what is harder to maintain).
In this case, maybe it would be better not having different function names for this. So what could be done is:
- Your function could be renamed to "calc_deflection_parameter" and be present in utils.
- The method "calc_deflection_parameter" in  the ´SinuosidalFieldSource´ object would just call your new function from utils and execute it. 
- A new method "calc_b_from_deflection_parameter" could do the inverse. And follow the same logic: a function in utils and a new method in `SinuosidalFieldSource` which internally calls the method in utils to perform the calculation.
I believe that if this suggestion implemented, changing the name of the new function is better than changing the name of the old one to match, since changing the old name would break our older scripts.

![classes_imaids](https://user-images.githubusercontent.com/61742723/186710772-00b50803-d25f-48eb-a62e-724dac8cb381.png)

Reply:
All right, so what I've done was:

- I renamed the function "undulator_b_to_k" to  "calc_deflection_parameter" and now the method "calc_deflection_parameter" from SinusoidalFieldSource just call this function - which is still on utils.py -  to execute the calculation.

I realized that the funtion calc_b_from_deflection_parameter wasn't being used anywhere, so i just deleted it.